### PR TITLE
feature(tj): Add monitoring for the gradient conflict metric of MoE.

### PR DESCRIPTION
**Pull request by tAnGjIa520**

### CODE REVIEW

File: lzero/entry/train_unizero_multitask_segment_ddp.py [hunk 1]
```diff
@@ -13,17 +13,35 @@
 from ding.worker import BaseLearner
 from tensorboardX import SummaryWriter
 
-from lzero.entry.utils import log_buffer_memory_usage, TemperatureScheduler
+# 添加性能监控相关导入
```
Comment (puyuan1996) [inline lzero/entry/train_unizero_multitask_segment_ddp.py:16]: 改一下注释为英文，统一用类似[DI-engine/ding/model/common/encoder.py at main · opendilab/DI-engine](https://github.com/opendilab/DI-engine/blob/main/ding/model/common/encoder.py)这样的规范格式

File: diff_output.txt [hunk 2]
```diff
@@ -0,0 +1,2207 @@
+diff --git a/lzero/entry/train_unizero_multitask_segment_ddp.py b/lzero/entry/train_unizero_multitask_segment_ddp.py
```
Comment (puyuan1996) [inline diff_output.txt:1]: 删除不用的文件

File: lzero/entry/train_unizero_multitask_segment_ddp.py [hunk 3]
```diff
@@ -13,17 +13,35 @@
...
+)
 from lzero.policy import visit_count_temperature
 from lzero.worker import MuZeroEvaluator as Evaluator
 from lzero.worker import MuZeroSegmentCollector as Collector
 from ding.utils import EasyTimer
 import torch.nn.functional as F
-
+import sys
+import os
+PROJECT_ROOT = os.path.abspath("/fs-computility/niuyazhe/tangjia/github/LightZero") # 或者直接写死路径
```
Comment (puyuan1996) [inline lzero/entry/train_unizero_multitask_segment_ddp.py:33]: 这个是必要的吗 如果是写成相对路径，不要带有个人信息

File: lzero/entry/train_unizero_multitask_segment_ddp.py [hunk 4]
```diff
@@ -37,8 +55,788 @@ def build_learner_group(learner_ranks: list[int]) -> dist.ProcessGroup:
     if dist.get_rank() in learner_ranks:
         torch.cuda.set_device(learner_ranks.index(dist.get_rank()))
     return pg
+
+
+# ------------------------------------------------------------
+# MOE专家选择统计相关函数
+# ------------------------------------------------------------
+def merge_expert_stats_across_ranks(all_expert_stats):
```
Comment (puyuan1996) [inline lzero/entry/train_unizero_multitask_segment_ddp.py:63]: 通用工具函数放在utils中去

File: lzero/entry/train_unizero_multitask_segment_ddp.py [hunk 5]
```diff
@@ -37,8 +55,788 @@ def build_learner_group(learner_ranks: list[int]) -> dist.ProcessGroup:
...
+    if _GLOBAL_HEATMAP_FIG is None:
+        _GLOBAL_HEATMAP_FIG, _GLOBAL_HEATMAP_AX = plt.subplots(figsize=figsize)
+    else:
+        # 清除之前的内容
+        _GLOBAL_HEATMAP_AX.clear()
+        # 调整图像大小
+        _GLOBAL_HEATMAP_FIG.set_size_inches(figsize)
+    return _GLOBAL_HEATMAP_FIG, _GLOBAL_HEATMAP_AX
+
+def create_heatmap_with_values_fast(matrix, task_ids, title="Task-Expert Selection Frequencies"):
```
Comment (puyuan1996) [inline lzero/entry/train_unizero_multitask_segment_ddp.py:100]: 通用工具函数放在utils中去 这里也是

File: lzero/entry/utils.py [hunk 6]
```diff
@@ -362,3 +363,1005 @@ def log_buffer_run_time(train_iter: int, buffer: "GameBuffer", writer: SummaryWr
...
+        'gpu_accelerated': 'cuda' in str(device),
+        
+        # 返回CPU版本用于记录
+        'js_matrix': js_matrix.cpu().numpy(),
+        'wasserstein_matrix': wasserstein_matrix.cpu().numpy(),
+        'js_stats': js_stats,
+        'wasserstein_stats': wasserstein_stats
+    }
+
+def create_similarity_heatmap_no_diagonal(similarity_matrix, task_ids, metric_name, title_suffix=""):
```
Comment (puyuan1996) [inline lzero/entry/utils.py:976]: 工具函数已经放在utils里面了，就把lzero/entry/train_unizero_multitask_segment_ddp.py里面没用的删除了哈

File: lzero/entry/utils.py [hunk 7]
```diff
@@ -362,3 +363,1005 @@ def log_buffer_run_time(train_iter: int, buffer: "GameBuffer", writer: SummaryWr
...
+        torch.Tensor: Wasserstein距离矩阵，形状 (n_tasks, n_tasks)，对称矩阵
+    
+    Note:
+        - 使用GPU优化器的内存池减少内存分配
+        - 通过CDF计算Wasserstein-1距离
+        - 支持向量化计算，高效处理大批量数据
+        - 主要用于衡量任务间专家选择的差异性
+    """
+    # 使用GPU优化器提升性能
+    return get_gpu_optimizer().optimized_wasserstein(distributions_tensor)
```
Comment (puyuan1996) [inline lzero/entry/utils.py:884]: 这里是为了gpu上计算散度性能更加高效吗？可以大致写一下相比基线速度提升了多少

File: lzero/model/common.py [hunk 8]
```diff
@@ -248,6 +248,68 @@ def remove_hooks(self):
         self.forward_handler.remove()
         self.backward_handler.remove()
 
+# # modified by tangjia
+# class ModelGradientHook:
```
Comment (puyuan1996) [inline lzero/model/common.py:252]: 删除不用的注释

File: lzero/model/unizero_model_multitask.py [hunk 9]
```diff
@@ -124,7 +124,7 @@ def __init__(
                     ))
             elif world_model_cfg.encoder_type == "vit":
                 for task_id in range(1):  # TODO: one share encoder
-                    if world_model_cfg.task_num <=8: 
+                    if world_model_cfg.task_num ==1:
```
Comment (puyuan1996) [inline lzero/model/unizero_model_multitask.py:127]: 为什么单独加一个task_num ==1的呢？

File: lzero/model/unizero_model_multitask.py [hunk 10]
```diff
@@ -189,6 +218,11 @@ def __init__(
                 self.encoder_hook = FeatureAndGradientHook()
                 self.encoder_hook.setup_hooks(self.representation_network)
 
+            # if True: # Fixme: for debug
+            #     # 增加对encoder的hook,监控传播到encoder 上的梯度
+            #     self.encoder_output_hook = ModelGradientHook()
```
Comment (puyuan1996) [inline lzero/model/unizero_model_multitask.py:223]: 删除 或者保留，增加标准英文注释

File: lzero/model/unizero_world_models/moe.py [hunk 11]
```diff
@@ -69,34 +70,54 @@ def __init__(self, config, experts: List[nn.Module], gate: nn.Module, num_expert
...
-            if batch_idx.numel() == 0:
-                continue
-            token_subset = x[batch_idx]  # 选中的 token，形状 [num_tokens, dim]
-            # 调用当前专家模块计算输出
-            output_expert = self.experts[expert_id](token_subset)
-            # 获取对应 token 的权重，注意 weights 的形状为 [N, num_experts_per_tok]
-            token_weights = weights[batch_idx, expert_tok_idx].unsqueeze(-1)
-            expert_output[batch_idx] += output_expert * token_weights
+        expert_output=x
+        if self.num_experts!=0:
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/moe.py:94]: 为什么增加self.num_experts!=0呢？self.num_experts不是一定大于0吗

File: lzero/entry/train_unizero_multitask_segment_ddp.py (hunk 1 above)
Comment (puyuan1996) [reply lzero/entry/train_unizero_multitask_segment_ddp.py:16]: 你只改相对dev-multitask-balance-clean分支的moe grad改动这块哈，dev-multitask-balance-clean分支内部的注释规范我去修改哈

File: lzero/model/unizero_world_models/transformer.py [hunk 12]
```diff
@@ -546,13 +730,14 @@ def __init__(self, config: TransformerConfig) -> None:
                 _maybe_wrap_linear(nn.Linear(config.embed_dim, 4 * config.embed_dim), config, "feed_forward"),
                 nn.GELU(approximate='tanh'),
                 _maybe_wrap_linear(nn.Linear(4 * config.embed_dim, config.embed_dim), config, "feed_forward"),
-                nn.Dropout(config.resid_pdrop),
+                # nn.Dropout(config.resid_pdrop),
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/transformer.py:733]: 为什么去掉了这个呢

File: lzero/model/unizero_world_models/transformer.py [hunk 13]
```diff
@@ -450,6 +462,178 @@ def forward(
             x = x[:, :-self.register_token_num, :]
 
         return x
+    
+    def get_expert_selection_stats(self, task_id: int = None):
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/transformer.py:466]: transformer里面与grad分析相关的改动，担心会影响到原来的性能，因为没有充分测试，暂时写成2个文件吧 一个是原来的 transformer.py 一个是改动后的 transformer_analysis.py 注释里面写一下

File: lzero/entry/train_unizero_multitask_segment_ddp.py [hunk 14]
```diff
@@ -810,7 +1630,71 @@ def train_unizero_multitask_segment_ddp(
 
                     # 在训练时，DDP会自动同步梯度和参数
                     log_vars = learner.train(train_data_multi_task, envstep_multi_task, policy_kwargs=learn_kwargs)
-
+                    
+                    print("训练结束！！！")
+                    
+                    # +++++++++++++++++++++++++++++++++ MOE专家选择统计记录 +++++++++++++++++++++++++++++++++
```
Comment (puyuan1996) [inline lzero/entry/train_unizero_multitask_segment_ddp.py:1636]: 不用的先删除了哈，如果后面debug要用到，你本地先backup一个分支就好

File: lzero/mcts/tree_search/mcts_ctree.py [hunk 15]
```diff
@@ -56,9 +56,13 @@ def __init__(self, cfg: EasyDict = None) -> None:
         default_config = self.default_config()
         default_config.update(cfg)
         self._cfg = default_config
+        if eval:
+            self._cfg.num_simulations=self._cfg.eval_num_simulations
```
Comment (puyuan1996) [inline lzero/mcts/tree_search/mcts_ctree.py:60]: eval之后self._cfg.num_simulations的值就变了，下次collect的num_simulations不是被覆盖了吗

File: lzero/policy/unizero_multitask.py [hunk 16]
```diff
@@ -418,7 +433,7 @@ def _init_learn(self) -> None:
             device_type=self._cfg.device,
             betas=(0.9, 0.95),
         )
-
+        # self.a=1
```
Comment (puyuan1996) [inline lzero/policy/unizero_multitask.py:436]: 删除

File: lzero/policy/unizero_multitask.py [hunk 17]
```diff
@@ -988,7 +1166,26 @@ def _monitor_vars_learn(self, num_tasks=2) -> List[str]:
...
+            # 'avg_shared_expert_grad_conflict',
+
+            # 'max_encoder_grad_conflict',
+            # 'max_before_moe_grad_conflict',
+            # 'max_shared_expert_grad_conflict',
+            # "avg_moe_layer_grad_conflict",
+            # "max_moe_layer_grad_conflict",
         ]
+        
+        # # # If the model uses MoE, add expert gradient conflict variables
```
Comment (puyuan1996) [inline lzero/policy/unizero_multitask.py:1181]: 删除不用的注释

File: zoo/atari/config/atari_unizero_multitask_segment_ddp_config_debug.py [hunk 18]
```diff
@@ -1,6 +1,11 @@
 from easydict import EasyDict
 
 import math
+import sys
+import os
+PROJECT_ROOT = os.path.abspath("/fs-computility/niuyazhe/tangjia/github/LightZero") # 或者直接写死路径
+sys.path.insert(0, PROJECT_ROOT)
+# /fs-computility/niuyazhe/tangjia/github/LightZero/zoo/atari/config/atari_unizero_multitask_segment_ddp_config_debug.py
```
Comment (puyuan1996) [inline zoo/atari/config/atari_unizero_multitask_segment_ddp_config_debug.py:8]: 删除debug config

File: zoo/atari/config/atari_unizero_multitask_segment_ddp_config_debug.py [hunk 19]
```diff
@@ -363,21 +371,28 @@ def create_env_manager():
...
         with DDPContext():
+            
+            # print(train_unizero_multitask_segment_ddp.__file__)
             train_unizero_multitask_segment_ddp(configs, seed=seed, max_env_step=max_env_step, benchmark_name= "atari" )
             # ======== TODO: only for debug ========
             # train_unizero_multitask_segment_ddp(configs[:2], seed=seed, max_env_step=max_env_step) # train on the first four tasks
         
-        # 手动销毁进程组
+        # 手动销毁进程组 /fs-computility/niuyazhe/tangjia/github/LightZero/zoo/atari/config/atari_unizero_multitask_segment_ddp_config_debug.py
         if dist.is_initialized():
```
Comment (puyuan1996) [inline zoo/atari/config/atari_unizero_multitask_segment_ddp_config_debug.py:397]: config中只保留我们scalezero论文里面用到分析的吧

File: lzero/mcts/tree_search/mcts_ctree.py (hunk 15 above)
Comment (tAnGjIa520) [reply lzero/mcts/tree_search/mcts_ctree.py:60]: 这里是初始化，相当于 我在初始化eval 的UniZeroMCTSCtree的时候，把里面的num_simulations 设置为我传入的eval_num_simulations
